homework_solution.md

```diff
@@ -0,0 +1,474 @@
+          end_time = (SELECT TO_TIMESTAMP(end_time_str, 'MM/DD/YY HH24:MI'))
+       WHERE original_filename LIKE '%2017 Q4%'
+       AND start_time_str IS NOT NULL
+       AND id <> 2302635;
+       ```
+
+       (While dealing with this, we found that record `2302635` has `end_time_str` of `NULLNULL`, so we'll treat it separately)
+
+       ```
+       UPDATE trips
+       SET
+          start_time = (SELECT TO_TIMESTAMP(start_time_str, 'MM/DD/YY HH24:MI')),
+          end_time_str = NULL
+       WHERE id = 2302635;
+       ```
+
+    5. Update `2018` records
+       ```
+       UPDATE trips
+       SET
+          start_time = (SELECT TO_TIMESTAMP(start_time_str, 'MM/DD/YYYY HH24:MI')),
+          end_time = (SELECT TO_TIMESTAMP(end_time_str, 'MM/DD/YYYY HH24:MI'))
+       WHERE original_filename LIKE '%2018%'
+       AND start_time_str IS NOT NULL;
+       ```
+
+3.  _**Other than the index in class, would we benefit from any other indexes on this table? Why or why not?**_
+
+    Yes. An index on `original_filename` goes a long way here.
+    It can be perceived by the observing the different `cost` before and after the index:
+
+    ```
+
+    postgres=# EXPLAIN SELECT original_filename, (ARRAY_AGG(DISTINCT start_time_str))[1]
+    FROM trips
+    GROUP BY original_filename;
+    QUERY PLAN
+
+    ---
+
+    GroupAggregate (cost=719052.22..744667.25 rows=8 width=69)
+    Group Key: original_filename
+    -> Sort (cost=719052.22..727590.53 rows=3415324 width=52)
+    Sort Key: original_filename
+    -> Seq Scan on trips (cost=0.00..114957.24 rows=3415324 width=52)
+    JIT:
+    Functions: 7
+    Options: Inlining true, Optimization true, Expressions true, Deforming true
+    (8 rows)
+
+    ```
+
+    ```
+
+    postgres=# CREATE INDEX idx_original_filename ON trips USING btree (original_filename);
+    CREATE INDEX
+
+    ```
+
+    ```
+
+    postgres=# EXPLAIN SELECT original_filename, (ARRAY_AGG(DISTINCT start_time_str))[1]
+    FROM trips
+    GROUP BY original_filename;
+    QUERY PLAN
+
+    ---
+
+    GroupAggregate (cost=0.56..343851.30 rows=8 width=69)
+    Group Key: original_filename
+    -> Index Scan using idx_original_filename on trips (cost=0.56..326774.58 rows=3415324 width=52)
+    JIT:
+    Functions: 5
+    Options: Inlining false, Optimization false, Expressions true, Deforming true
+    (6 rows)
+
+    ```
+
+# Part 3: Data-driven insights
+
+1. _**Build a mini-report that does a breakdown of number of trips by month**_
+
+   ```
+   SELECT
+      CASE  EXTRACT (MONTH FROM start_time)
+         WHEN 1 THEN '01 - JAN'
+         WHEN 2 THEN '02 - FEB'
+         WHEN 3 THEN '03 - MAR'
+         WHEN 4 THEN '04 - APR'
+         WHEN 5 THEN '05 - MAY'
+         WHEN 6 THEN '06 - JUN'
+         WHEN 7 THEN '07 - JUL'
+         WHEN 8 THEN '08 - AUG'
+         WHEN 9 THEN '09 - SEP'
+         WHEN 10 THEN '10 - OCT'
+         WHEN 11 THEN '11 - NOV'
+         WHEN 12 THEN '12 - DEC'
+      END
+      AS Month,
```

Comment:
[question]
Why denormalize the month names?  Can you think of any other ways to accomplish that goal?

Reply:
Just because it was called a **report** and to me having the month names make it more readable.
Also, I don't know of other ways in SQL to get that.

Reply:
That's a good reason - it's important to think about the final use case.  It doesn't matter how well-engineered something is if it doesn't get the job done!

Another approach I was thinking about (which is not necessarily a good idea, just a possibility) would be to define a `months` table which maps month numbers to their name, and then do a `JOIN` on that in the final report.  In my experience, reports usually aren't generated with pure SQL though - there's often another application on top of the database that formats it nicely (e.g. a python script, some database front end etc), in which case you'd want to keep all of the denormalization/humanization of data in that higher application/presentation layer.

Reply:
Oh, for sure about other layers for properly built reports. In this case, having another table feels to me like a possible overkill (and maybe more costly to run the `JOIN` then the `CASES`, but not sure).
I think I'd go for another table where either:
a. There'd be more than a single value I'd like to show based on the same value (ie. `if Month = 1 then Column A = 'Jan', Column B = 'January', etc)
b. The actual values to be presented based on an id are more likely to change (ie. if there was locale to consider and we had to retrieve the correct string for a specific location)